server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java

```diff
@@ -112,6 +115,106 @@ static boolean isNotUnitVector(float magnitude) {
     public static short MIN_DIMS_FOR_DYNAMIC_FLOAT_MAPPING = 128; // minimum number of dims for floats to be dynamically mapped to vector
     public static final int MAGNITUDE_BYTES = 4;
 
+    interface VectorByteDataConsumer {
+        void accept(byte value, int index);
+    }
+
+    public static class VectorData {
```

Comment:
Not sure if this is the correct place or not - but thought to add it here initially, as almost all related classes already had a dependency on `DenseVectorFieldMapper` - happy to move to its own class / discuss alternatives.

Reply:
- I think it should be its own top-level class in `org.elasticsearch.search.vectors`.
 - It should be a `record`, you can override the canonical for uniqueness checks
 - It should also override `Writeable` and possibly `ToXContent` and handle all the serialization stuff.
 - It should handle its own XContent parsing, this way users of this only need to use this class and it can correctly parse numerical array or string values

Reply:
++  Tbh the main reason that I defined it as a `class` instead of a `record` was to hide constructors and enable new object creation only via static methods (to ensure uniqueness).

Reply:
> was to hide constructors and enable new object creation only via static methods (to ensure uniqueness).

You can do that in the `canonical` ctor and still have static methods that are preferred. 

I would do something like:

```
record VectorData(float[] floats, byte[] bytes) {
    public VectorData {
        if (floats == null ^ bytes == null) {
          throw new IllegalArgumentException("You must supply exactly either floats or bytes");
        }
    }
}
```